### PR TITLE
RSDK-142 change gantry oneaxis structure

### PR DESCRIPTION
**Pull request by randhid**
Changes the one-axis structure to assume that we have knowledge of the position of gantry limits, and wraps that in limit switch logic if that is available. Both still need to "home", but axes with encoded motors can do so using math based on motor position and length input by the user.

Added tests for more coverage, and code cleanup. Hardware re-testing is needed when I am back, but not blocking.
- [ ]  Test on hardware when back in the office.

### CODE REVIEW

File: components/gantry/oneaxis/setup.go [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+			return
+		default:
+		}
+		for idx := range g.limitSwitchPins {
+			hit, err := g.limitHit(g.ctx, idx)
+			if hit || err != nil {
+				g.limitHitMu.Lock()
+				g.limitsPolled = false
+				g.limitHitMu.Unlock()
+				errs := multierr.Combine(errs, err)
```
Comment (penguinland) [inline components/gantry/oneaxis/setup.go:230]: This line seems unnecessary: we know `errs` is nil because it's nil the first time through and every iteration except the first one, we would have returned at the end of the previous iteration if it was not nil.

So, I think the 4 lines starting here can all be condensed:
```go
// motor operation manager cancels running
errs = multierr.Combine(err, g.oAx.motor.Stop(g.ctx, nil))
```

Comment (penguinland) [reply components/gantry/oneaxis/setup.go:230]: The `:=` operator creates a brand new variable `errs` which shadows the one created on line 217. I suspect line 217 should be removed, though I could be wrong about that.

File: components/gantry/oneaxis/setup.go [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+					return
+				}
+			}
+		}
+	}
+}
+
+func (g *limitSwitchGantry) limitHit(ctx context.Context, offset int) (bool, error) {
+	if offset < 0 || offset > 1 {
+		return true, errBadNumLimitSwitches(g.oAx.name, offset)
```
Comment (penguinland) [inline components/gantry/oneaxis/setup.go:245]: If `offset` is 2, we're going to get a confusing error message about how it can be 0, 1, or 2 but not 2. I suspect you're returning the wrong error here, though I'm unsure what the right one is. or maybe the if statement should check `offset > 2` instead?

File: components/gantry/oneaxis/setup.go [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+	pin, err := g.oAx.board.GPIOPinByName(g.limitSwitchPins[offset])
+	if err != nil {
+		return false, err
+	}
+	high, err := pin.Get(ctx, nil)
+
+	return high == g.limitHigh, err
+}
+
+func (g *limitSwitchGantry) testLimit(ctx context.Context, limInt int) (float64, error) {
```
Comment (penguinland) [inline components/gantry/oneaxis/setup.go:256]: What is the float that this function returns? Put in a quick comment explaining.

File: components/gantry/oneaxis/setup.go [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+		}
+		if hit {
+			err = g.oAx.motor.Stop(ctx, nil)
+			if err != nil {
+				return 0, err
+			}
+			break
+		}
+
+		elapsed := start.Sub(start)
```
Comment (penguinland) [inline components/gantry/oneaxis/setup.go:285]: Isn't `start.Sub(start)` always going to be 0? I wonder if you meant `time.Now().Sub(start)` instead.

File: components/gantry/oneaxis/setup.go [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+		if elapsed > (time.Second * 15) {
+			return 0, errors.New("gantry timed out testing limit")
```
Comment (penguinland) [inline components/gantry/oneaxis/setup.go:287]: This is probably good enough for now, but I can imagine gantries that take more than 15 seconds to home, which will cause trouble here. Maybe put in a TODO statement or similar to make this configurable in the future? Maybe it's fine as-is.

File: components/gantry/oneaxis/setup.go [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+		}
+
+		if !utils.SelectContextOrWait(ctx, time.Millisecond*10) {
```
Comment (penguinland) [inline components/gantry/oneaxis/setup.go:290]: This is basically a busy loop with a small sleep in it to make it not-quite-so-busy. It feels like what we really want is for the limit switch to be a digital interrupt instead of a GPIO pin, so that we can replace the for loop with a select statement waiting on either the context to be canceled, the timeout to occur, or the limit switch to get hit:
```
timeout := time.NewTicker(15 * Time.Second)
defer timeout.Stop()
g.oAx.motor.GoFor(...)
defer g.oAx.motor.Stop(...)

select {
    case <-g.ctx.Done():
        return 0, ctx.Err()
    case <-timeout.C:
        return 0, errors.New("gantry timed out")
    case <-g.limitSwitchHit:
        return g.oAx.motor.Position(...)
}
```
but Gautham's recent work with the ultrasonic sensor makes me question whether digital interrupts have decent performance at all, so maybe this isn't a great approach.

File: components/gantry/oneaxis/setup.go [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
+		}
+		revPerLength := g.oAx.lengthMm / g.oAx.mmPerRevolution
+		positionB = positionA + revPerLength
+	case 2:
+		positionB, err = g.testLimit(ctx, 1)
+		if err != nil {
+			return err
+		}
+	default:
+		return errBadNumLimitSwitches(g.oAx.name, np)
```
Comment (penguinland) [inline components/gantry/oneaxis/setup.go:321]: If someone passes in the empty list for `limSwitchPins`, they're going to get a confusing error about how the number of limit switches can be 0, 1, or 2 but not 0. I think this should be its own custom error message, though I don't feel strongly.

File: components/gantry/oneaxis/setup.go [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+	}
+
+	np := len(g.limitSwitchPins)
+	var positionB float64
+	switch len(limSwitchPins) {
+	case 1:
+		if g.oAx.lengthMm == 0.0 || g.oAx.mmPerRevolution == 0.0 {
+			return errDimensionsNotFound(g.oAx.name, g.oAx.lengthMm, g.oAx.mmPerRevolution)
+		}
+		revPerLength := g.oAx.lengthMm / g.oAx.mmPerRevolution
```
Comment (penguinland) [inline components/gantry/oneaxis/setup.go:313]: Mild preference to rename this `totalRevolutions` or similar because its units are revolutions and not revolutions per meter. but I don't feel strongly.

File: components/gantry/oneaxis/setup.go [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+
+func (g *limitSwitchGantry) homeWithLimSwitch(ctx context.Context, limSwitchPins []string) error {
+	ctx, done := g.oAx.opMgr.New(ctx)
+	defer done()
+	positionA, err := g.testLimit(ctx, 0)
+	if err != nil {
+		return err
+	}
+
+	np := len(g.limitSwitchPins)
```
Comment (penguinland) [inline components/gantry/oneaxis/setup.go:306]: This is used on line 321 and nowhere else. I'd prefer to move it down there; having it up here implies that it needs to have a much larger scope than it actually does.

File: components/gantry/oneaxis/setup.go [hunk 10]
```diff
@@ -0,0 +1,390 @@
...
+	case 2:
+		positionB, err = g.testLimit(ctx, 1)
+		if err != nil {
+			return err
+		}
+	default:
+		return errBadNumLimitSwitches(g.oAx.name, np)
+	}
+
+	g.oAx.logger.Debugf("positionA: %0.2f positionB: %0.2f", positionA, positionB)
```
Comment (penguinland) [inline components/gantry/oneaxis/setup.go:324]: Slight preference to add some more contextual words here: `Debugf("finished homing gantry. positionA: %0.2f...")` I get disoriented when I'm looking through the logs and there's data without context in there.

File: components/gantry/oneaxis/setup.go [hunk 11]
```diff
@@ -0,0 +1,390 @@
...
+
+	g.oAx.positionLimits = []float64{positionA, positionB}
+
+	if g.oAx.gantryRange = positionB - positionA; g.oAx.gantryRange == 0 {
+		return errZeroLengthGantry
+	}
+
+	// Go backwards so limit stops are not hit.
+	x := g.oAx.linearToRotational(0.9 * (g.oAx.gantryRange + positionA))
```
Comment (penguinland) [inline components/gantry/oneaxis/setup.go:333]: If your gantry has 2 limit switches, this makes perfect sense: we're near the far end, so just back off 10%. but if it only has 1 limit switch, this is going to travel nearly the entire range from the near end. Mild preference to make this slightly more intelligent and go 10% from whichever end it's nearest.

```go
target := 0.9
if len(limSwitchPins) == 1 {
    target = 0.1  // We're near the start, not the end
}
x := g.oAx.linearToRotational(target * g.oAx.gantryRange + positionA)

File: components/gantry/oneaxis/setup.go [hunk 12]
```diff
@@ -0,0 +1,390 @@
...
+}
+
+func (g *limitSwitchGantry) ModelFrame() referenceframe.Model {
+	g.mu.Lock()
+	defer g.mu.Unlock()
+	return g.oAx.ModelFrame()
+}
+
+func (g *limitSwitchGantry) CurrentInputs(ctx context.Context) ([]referenceframe.Input, error) {
+	return g.oAx.CurrentInputs(ctx)
```
Comment (penguinland) [inline components/gantry/oneaxis/setup.go:377]: Naive question: should this lock the mutex too? I don't know what `CurrentInputs` does; maybe the answer is obvious if you know that.

File: components/gantry/oneaxis/setup.go [hunk 13]
```diff
@@ -0,0 +1,390 @@
...
+	err = g.oAx.motor.GoTo(ctx, g.oAx.rpm, x, nil)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (g *limitSwitchGantry) Position(ctx context.Context, extra map[string]interface{}) ([]float64, error) {
+	g.mu.Lock()
+	defer g.mu.Unlock()
```
Comment (penguinland) [inline components/gantry/oneaxis/setup.go:343]: `g.mu` is only locked for these "passthrough" functions that just call the same thing on `g.oAx`. I'm not convinced we need this mutex: I'd expect that each function in `g.oAx` should be atomic, and thus not need a lock out here. If `g.mu` should exist, I suspect it should also be locked elsewhere in this file, but I don't see where that should be.

File: components/gantry/oneaxis/setup.go (hunk 11 above)
Comment (penguinland) [reply components/gantry/oneaxis/setup.go:333]: Oof, I missed this at first! Remove the innermost parentheses: the 0.9 should only multiply the range, not the starting position.

As an example of trouble, imagine that positionA is a million and gantryRange is a hundred.

**Review by penguinland — COMMENTED** [review]
So far I've only looked through setup.go. I'll get to the other files soon, hopefully...

File: components/gantry/oneaxis/setup.go (hunk 6 above)
Comment (penguinland) [reply components/gantry/oneaxis/setup.go:290]: No, I take that last part back: Gautham's work required microsecond responses, whereas with this motor it's good enough if you're off by a whole millisecond. I think digital interrupts are likely the way to go when implementing limit switches.

File: components/gantry/oneaxis/setup.go [hunk 14]
```diff
@@ -0,0 +1,390 @@
...
+	if cfg.LengthMm <= 0 {
+		return nil, errors.New("each axis needs a non-zero and positive length")
+	}
+
+	if len(cfg.LimitSwitchPins) == 0 && len(cfg.Board) > 0 {
+		return nil, errors.New("gantry with encoders have to assign boards or controllers to motors")
+	}
+
+	if len(cfg.Board) == 0 && len(cfg.LimitSwitchPins) > 0 {
+		return nil, errors.New("cannot find board for gantry")
```
Comment (gvaradarajan) [inline components/gantry/oneaxis/setup.go:69]: could we include a little more detail in this error message? since you only need to supply a board if you're using a limit switch gantry, this message could be misleading to a user (I like the specificity you used in L. 65, for example)

File: components/gantry/oneaxis/setup.go (hunk 12 above)
Comment (gvaradarajan) [reply components/gantry/oneaxis/setup.go:377]: I agree with this and the previous comment. Either there should be a mutex on the one axis struct that locks on the interior functions and these outer locks should be removed, or this function should make use of the mutex on `g`. The former seems more scalable to me, but I don't know how much of a concern scalability is in the case of gantries so whichever you think is clearer is probably fine

**Review by gvaradarajan — CHANGES_REQUESTED** [review]
I only had a couple additional comments to add to the review already left by @penguinland .

File: components/gantry/oneaxis/setup.go (hunk 6 above)
Comment (gvaradarajan) [reply components/gantry/oneaxis/setup.go:290]: Upon closer examination, we found the digital interrupt latency to not be an issue unless you wanted the specific time that a tick was received (and even this problem is being resolved). It shouldn't be an issue here should you use a digital interrupt

**Review by penguinland — COMMENTED** [review]
I'm having a very hard time reading the tests: it's like you've removed all of the old tests and added a bunch of new tests, and github's diffs are a mess of coincidences between the two. I'm going to stop reviewing the tests, and will trust that if they pass, you're probably doing good things in them. The non-test code is the important part anyway.

File: components/gantry/oneaxis/oneaxis.go [hunk 15]
```diff
@@ -160,130 +68,67 @@ func newOneAxis(ctx context.Context, deps registry.Dependencies, config config.C
...
-		err := g.homeEncoder(ctx)
-		if err != nil {
-			return err
-		}
-	}
-	return nil
-}
+	f, err := referenceframe.NewStaticFrame(g.name, spatialmath.NewZeroPose())
+	errs = multierr.Combine(errs, err)
+	m.OrdTransforms = append(m.OrdTransforms, f)
```
Comment (penguinland) [inline components/gantry/oneaxis/oneaxis.go:102]: Naive question: if there are errors early on, why continue to accumulate them, rather than returning early?

File: components/gantry/oneaxis/oneaxis.go [hunk 16]
```diff
@@ -160,130 +68,67 @@ func newOneAxis(ctx context.Context, deps registry.Dependencies, config config.C
...
-	revPerLength := g.lengthMm / g.mmPerRevolution
-	positionB := positionA + revPerLength
-
-	g.positionLimits = []float64{positionA, positionB}
-
-	return nil
+func (g *oneAxis) linearToRotational(positions float64) float64 {
+	g.mu.Lock()
+	defer g.mu.Unlock()
+	theRange := g.positionLimits[1] - g.positionLimits[0]
```
Comment (penguinland) [inline components/gantry/oneaxis/oneaxis.go:119]: Throughout this PR, we're assuming the positions are sorted smallest to biggest. That's probably a good assumption, but I wonder if we should put a comment somewhere explicitly pointing it out. Should it go in the homing function? or at the top of a file as a package-level comment? or maybe not at all. but it seems to me that anyone who hooks their encoder up backwards is going to have a bad time, because the first position will be larger than the second.

File: components/gantry/oneaxis/oneaxis.go [hunk 17]
```diff
@@ -294,89 +139,32 @@ func (g *oneAxis) homeEncoder(ctx context.Context) error {
...
+	defer g.mu.Unlock()
 	pos, err := g.motor.Position(ctx, extra)
 	if err != nil {
 		return []float64{}, err
 	}
 
-	theRange := g.positionLimits[1] - g.positionLimits[0]
-	x := g.lengthMm * ((pos - g.positionLimits[0]) / theRange)
-
+	// never divide by zero since we cannot create a gantry with a zero range
```
Comment (penguinland) [inline components/gantry/oneaxis/oneaxis.go:159]: This comment looks like a warning, but I think it's supposed to be a reassurance. I'd rephrase to "This will never divide by zero because..."

File: components/gantry/oneaxis/oneaxis.go [hunk 18]
```diff
@@ -294,89 +139,32 @@ func (g *oneAxis) homeEncoder(ctx context.Context) error {
...
+	x := g.lengthMm * ((pos - g.positionLimits[0]) / g.gantryRange)
 	return []float64{x}, nil
 }
 
 // Lengths returns the physical lengths of an axis of a Gantry.
 func (g *oneAxis) Lengths(ctx context.Context, extra map[string]interface{}) ([]float64, error) {
+	g.mu.Lock()
+	defer g.mu.Unlock()
```
Comment (penguinland) [inline components/gantry/oneaxis/oneaxis.go:167]: I doubt locking the mutex will hurt anything, but how could it help anything, either? I'd sure expect that everything used in this function could be considered immutable, in which case there are no race conditions to avoid.

File: components/gantry/oneaxis/oneaxis.go [hunk 19]
```diff
@@ -294,89 +139,32 @@ func (g *oneAxis) homeEncoder(ctx context.Context) error {
...
 	return []float64{g.lengthMm}, nil
 }
```
Comment (penguinland) [inline components/gantry/oneaxis/oneaxis.go:170]: Github won't let me comment on the code below here, but if you're looking for places to lock a mutex, I'd expect `MoveToPosition` to be a good one.

File: components/gantry/oneaxis/oneaxis.go [hunk 20]
```diff
@@ -449,25 +210,8 @@ func (g *oneAxis) IsMoving(ctx context.Context) (bool, error) {
...
-		if errs != nil {
-			g.logger.Error(errs)
-			return nil
-		}
-
-		m.OrdTransforms = append(m.OrdTransforms, f)
-		g.model = m
-	}
+	g.mu.Lock()
+	defer g.mu.Unlock()
```
Comment (penguinland) [inline components/gantry/oneaxis/oneaxis.go:214]: Again, I don't see how this mutex helps anything: we're immediately returning, so there's no chance of some other goroutine trying to do something in between two of the steps in this function. (The mutex certainly doesn't hurt: feel free to keep it this way if it gives peace of mind.)

File: components/gantry/oneaxis/oneaxis.go [hunk 21]
```diff
@@ -439,6 +199,7 @@ func (g *oneAxis) MoveToPosition(
 func (g *oneAxis) Stop(ctx context.Context, extra map[string]interface{}) error {
 	ctx, done := g.opMgr.New(ctx)
 	defer done()
+	g.opMgr.CancelRunning(ctx)
 	return g.motor.Stop(ctx, extra)
```
Comment (penguinland) [inline components/gantry/oneaxis/oneaxis.go:203]: Either lock the mutex in this function, or add a comment explaining that locking the mutex here will cause us to block until all movement is complete, thus defeating the purpose of `Stop` entirely.

If the latter, what happens if some other goroutine tries to interact with the motor after we've canceled the opMgr but before we've stopped the motor?

**Review by penguinland — COMMENTED** [review]
I've now read through all non-test code, and things look pretty good!

File: components/gantry/oneaxis/setup.go [hunk 22]
```diff
@@ -0,0 +1,390 @@
...
+func (cfg *AttrConfig) Validate(path string) ([]string, error) {
+	var deps []string
+
+	if len(cfg.Motor) == 0 {
+		return nil, errors.New("cannot find motor for gantry")
+	}
+	deps = append(deps, cfg.Motor)
+
+	if cfg.LengthMm <= 0 {
+		return nil, errors.New("each axis needs a non-zero and positive length")
```
[the remark this answers is not in the document]
Comment (randhid) [reply components/gantry/oneaxis/setup.go:61]: each single axis is a component that can be used by multi-axis to make that component. I'll change it to `a one-axis gantry...`

File: components/gantry/oneaxis/setup.go (hunk 11 above)
Comment (randhid) [reply components/gantry/oneaxis/setup.go:333]: really nice catch, thank you.

File: components/gantry/oneaxis/setup.go (hunk 12 above)
Comment (randhid) [reply components/gantry/oneaxis/setup.go:377]: I'll remove the outer locks, and add them to the interior functions where appropriate.

File: components/gantry/oneaxis/oneaxis_test.go [hunk 23]
```diff
@@ -163,81 +130,62 @@ func TestValidate(t *testing.T) {
...
-			GantryRPM:       float64(300),
-		},
-	}
-	fakegantry, err = newOneAxis(ctx, deps, fakecfg, logger)
-	fakeoneax, ok = fakegantry.(*oneAxis)
-	test.That(t, ok, test.ShouldBeTrue)
-	test.That(t, err, test.ShouldBeNil)
-	test.That(t, fakeoneax.limitType, test.ShouldEqual, "onePinOneLength")
+	fakecfg.ConvertedAttributes.(*AttrConfig).LimitPinEnabled = &setTrue
+	fakecfg.ConvertedAttributes.(*AttrConfig).LimitSwitchPins = []string{"1", "2"}
```
[the remark this answers is not in the document]
Comment (randhid) [reply components/gantry/oneaxis/oneaxis_test.go:158]: Was trying to set up the config once, and then change the attributes to make sure our constructor errors were being fed back. THe difference with the motor injection is that that motor was being created everywhere. I can try to find a better way to show this.

File: components/gantry/oneaxis/setup.go [hunk 24]
```diff
@@ -0,0 +1,390 @@
...
+		return nil, err
+	}
+
+	conf, ok := cfg.ConvertedAttributes.(*AttrConfig)
+	if !ok {
+		return nil, rdkutils.NewUnexpectedTypeError(conf, cfg.ConvertedAttributes)
+	}
+
+	oAx, ok := g.(*oneAxis)
+	if !ok {
```
[the remark this answers is not in the document]
Comment (randhid) [reply components/gantry/oneaxis/setup.go:127]: I need a bit more points to be addressed from this comment.

- `newOneAxis` returns a `gantry.Local`, I cast it again to be able to access `homeEncoder` in the switch statement below, to determine homing logic. You're saying, just return an instance of the struct for oAx, and just return the local gantry from this function?

File: components/gantry/oneaxis/setup.go (hunk 5 above)
Comment (randhid) [reply components/gantry/oneaxis/setup.go:279]: I think we have enough speed params from the motor and dimension params from the gantry that we could maybe parameterize it rather than configure it.

File: components/gantry/oneaxis/oneaxis.go (hunk 16 above)
Comment (randhid) [reply components/gantry/oneaxis/oneaxis.go:119]: made `g.theRange` and absolute value, which should allow us a correct range without caring about directionality. We will still always go backwards first. That would be a motor fix.

File: components/gantry/oneaxis/setup.go [hunk 25]
```diff
@@ -0,0 +1,390 @@
...
+
+	return g, nil
+}
+
+func (g *limitSwitchGantry) limitHitChecker() {
+	// lock the mutex associated with this monitoring function
+	g.limitHitMu.Lock()
+	// check if the state is false
+	if !g.limitsPolled {
+		return
```
[the remark this answers is not in the document]
Comment (randhid) [reply components/gantry/oneaxis/setup.go:202]: yes, thank you.

File: components/gantry/oneaxis/setup.go (hunk 4 above)
Comment (randhid) [reply components/gantry/oneaxis/setup.go:277]: good catch, thank you.

File: components/gantry/oneaxis/setup.go (hunk 7 above)
Comment (randhid) [reply components/gantry/oneaxis/setup.go:313]: Then that makes an encoded gantry only and never hits this code because of the setup function.

File: components/gantry/oneaxis/setup.go [hunk 26]
```diff
@@ -0,0 +1,390 @@
...
+	g := &limitSwitchGantry{
+		ctx:             ctx,
+		limitSwitchPins: conf.LimitSwitchPins,
+		limitHigh:       *conf.LimitPinEnabled,
+		oAx:             oneAxis,
+		logger:          logger,
+		limitsPolled:    false,
+	}
+
+	if len(g.limitSwitchPins) == 1 && (g.oAx.mmPerRevolution == 0 || g.oAx.lengthMm == 0) {
```
[the remark this answers is not in the document]
Comment (randhid) [reply components/gantry/oneaxis/setup.go:187]: `length_mm` is always a required field in `Validate`

File: components/gantry/oneaxis/oneaxis.go (hunk 15 above)
Comment (randhid) [reply components/gantry/oneaxis/oneaxis.go:102]: We'd like to get as much information on a reference frame out as possible. So that motion can use the pertinent information it can if it needs.

File: components/gantry/oneaxis/oneaxis.go (hunk 21 above)
Comment (randhid) [reply components/gantry/oneaxis/oneaxis.go]: I actually changed Stop back. motor's Stop implements the cancel Running anyway.

File: components/gantry/oneaxis/oneaxis.go (hunk 19 above)
Comment (randhid) [reply components/gantry/oneaxis/oneaxis.go]: Added, please let me know if you see more places to add mutexes.
